File: csharp/lesson 5/TopKFrequentWords.cs
// https://leetcode.com/problems/top-k-frequent-words

public class Solution
{
    public IList<string> TopKFrequent(string[] words, int k)
    {
        var freq = new Dictionary<string, int>();
        foreach (var s in words)
        {
            freq[s] = freq.GetValueOrDefault(s, 0) + 1;
        }

        var pq = new PriorityQueue<string, KeyValuePair<string, int>>(Comparer<KeyValuePair<string, int>>.Create((x, y) =>
                        (x.Value == y.Value) ? (y.Key.CompareTo(x.Key)) : (x.Value - y.Value)));

        foreach (var entry in freq)
        {
            pq.Enqueue(entry.Key, entry);
            if (pq.Count > k)
            {
                pq.Dequeue();
            }
        }

        var result = new string[pq.Count];
        for (var i = result.Count - 1; i >= 0; i--)
        {
            result[i] = pq.Dequeue();
        }
        return result;
    }
}

File: python/lesson 5/top_k_frequent_words.py
import collections
from heapq import heappop, heappush
from typing import List

# https://leetcode.com/problems/kth-largest-element-in-an-array


class Node:
    def __init__(self, word, freq):
        self.word = word
        self.freq = freq

    def __lt__(self, other):
        if self.freq == other.freq:
            return self.word > other.word
        return self.freq < other.freq


class Solution:
    def topKFrequent(self, words: List[str], k: int) -> List[str]:
        freq = collections.Counter(words)
        heap = list()
        for word, count in freq.items():
            node = Node(word, count)
            heappush(heap, node)
            if len(heap) > k:
                heappop(heap)

        result = list()
        while heap:
            result.append(heappop(heap).word)

        return result[::-1]
